src/wiktextract/extractor/es/models.py

```diff
@@ -0,0 +1,113 @@
+from typing import Optional
+import json
+
+import logging
+
+from pydantic import BaseModel, Field, model_validator
+from pydantic.json_schema import GenerateJsonSchema
+
+from wiktextract.wxr_context import WiktextractContext
+
+
+class PydanticLogger:
+    wxr: Optional[WiktextractContext] = None
+
+    @classmethod
+    def debug(
+        cls, msg: str, trace: Optional[str] = None, sortid: str = "XYZunsorted"
+    ):
+        if cls.wxr:
+            cls.wxr.wtp.debug(msg, trace=trace, sortid=sortid)
+        else:
+            logging.debug(msg)
+
+
+class BaseModelWrap(BaseModel):
+    class Config:
+        extra = "ignore"
```

Comment:
If any keys are added to a pydantic class that are not defined in the class definition, they will be ignored.

Reply:
This shouldn't be ignored, many classes in the current code already has `__slots__`. And `LoggingExtraFieldsModel` will also not needed.

Reply:
I am not quite sure what's the significance of `__slots__` in this context but let me explain my reasoning for configuring the base model to ignore extra fields.`

I think there are two goals in this context:
1. The pydantic models should describe the output accurately.
2. We want to capture as much data as possible.

In the German extractor, for example, I used the following pattern to extract the references to an example from a wikicode template:
```
for key, value in template_node.template_parameters.items():
    if isinstance(key, str):
        reference_data[key.lower()] = clean_node(wxr, {}, value)
```
In this example, the expected keys aren't defined anywhere in our code but rely implicitly on the arguments of that particular template. In this particular case, that makes a lot of sense since the reference template is quite detailed and exactly what we would like to capture.

However, when defining the json schema or using pydantic without the option `extra = "ignore"`, there is always the risk that our output has keys that aren't documented anywhere. That might not be a problem in and of itself but certainly makes this data a lot less valuable. We also wouldn't satisfy goal 1.

With the option `extra = "ignore"`, only keys that are defined in the pydantic model will be output and all others suppressed. This does satisfy goal 1.

However, we might still like to capture these keys (goal 2). For this reason, I am logging all extra fields that our code tries to assign but aren't defined in the model. This will enable us to make a decision on whether or not to include this key in our model. 

In short: The combination of ignoring extra fields while logging these in a debug statement will force us to keep the pydantic models (and thus our output documentation) in sync with what we do in our code. I think this is the main benefit of using pydantic over manually maintaining json schemas (as currently done for the French extractor).

Reply:
Thanks for the thorough review and pointing my in the right direction!

> The default value of extra is [ignore](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.extra), maybe the base model should use **forbid.**

With the option **forbid**, we would get an exception every time there is an extra field. This would mean that we need to add error handling every time we instantiate a pydantic model. I think that would be extremely cumbersome and I didn't find a good way to catch this exception in an abstract way... There is an interesting PR suggesting a fourth option [**warn**](https://github.com/pydantic/pydantic/issues/4945) which would be ideal for our case.

> Isn't the JSON schema file generated from the defined model classes? How could they contain fields created dynamically at runtime?

You're right. The JSON schema file will only contain fields defined in the model classes.

> As for your German data example, I think we should not check the fields(or generate JSON schema) not created in our code.

The main question is thus: Do we want that our JSON schemas strictly describe our output, i.e. the output contains only keys that are described in our JSON schemas? Or do we not mind if the output contains extra fields?

My preference and also the PR's is the first option.

Reply:
Sorry I missed this comment. If the code is initializing a pydantic model with undefined fields and get exception, that means the code or the model is wrong and that's the whole point of validation, right? The only way we could get this exception is from our own mistake but not from the changes in Wiktionary.